### ISSUE TITLE
feature request: allow workspace looping with repeatedly pressed keyboard shortcuts in workspace switcher applet when wrap around is enabled

### ISSUE THREAD

**Issue by N0rbert** [mate-desktop/mate-panel#969]
#### Expected behaviour
Workspace switching is done in a loop (wrap around) by one of the following methods:

* rotating mouse wheel above Workspace Switcher applet 
* interpreting keypresses on <Ctrl+Alt+ → > repeatedly (or <Ctrl+Alt+ ← > )

#### Actual behaviour
Workspace switching is done in a loop only by mouse.
Pressing <Ctrl+Alt+ → > repeatedly (or <Ctrl+Alt+ ← > ) moves to the last or to the first workspace.

#### Steps to reproduce the behaviour
1. Right click on Workspace Switcher,  select Preferences, check **[v] Allow workspace wrap around in switcher** (or run `dconf write /org/mate/panel/objects/workspace-switcher/prefs/wrap-workspaces 'true'`)
1. Press <Ctrl+Alt+ → > repeatedly (or <Ctrl+Alt+ ← > )

#### MATE general version
1.22.1 (but really it exists since MATE 1.8.1 as in Ubuntu 12.04 LTS with [MATE for Precise PPA](https://launchpad.net/~ubuntu-mate-dev/+archive/ubuntu/precise-mate))

#### Package version
1.22.1-0ubuntu1

#### Linux Distribution
Ubuntu 19.10

#### Link to downstream report of your Distribution
This is long story, ended here - https://bugs.launchpad.net/ubuntu-mate/+bug/1828842 .

**Comment by vkareh** [comment]
The best way to address this would be to synchronize the toggle in the workspace-switcher (`org.mate.panel.applet.workspace-switcher wrap-workspaces`) with the one in marco (`org.mate.Marco.general wrap-style`).

That means changing the checkbox to either a radio or a dropdown with the options for:
* No wrap
* Classic
* Toroidal

The behavior when scrolling on the applet would also need to be modified to support the Classic versus Toroidal options.

Then we can just get rid of `org.mate.panel.applet.workspace-switcher wrap-workspaces`, as it would be redundant.

**Comment by knopancakes** [comment]
@vkareh I like how you're thinking, consolidating the settings into a drop down. This should make the workspace settings feel less fragmented.

**Comment by vkareh** [comment]
The main question is what to do with Compiz/Metacity... we could disable it for non-marco window managers, as Compiz has its own UI for configuration...

the other alternative is to get rid of this toggle in the applet and add it instead to the `mate-window-properties`...

**Comment by knopancakes** [comment]
To be honest, I was surprised not to find this setting under mate-window-properties or mate-tweak when I first went looking. Perhaps mate-tweak would be a better candidate since it also has context for selecting the window manager?

**Comment by knopancakes** [comment]
But now that's going to open another can of worms, there being duplicate settings in mate-tweak/Windows and mate-window-properties

**Comment by vkareh** [comment]
possibly... you could put a request there. However, mate-tweak for now is Ubuntu-only... Fedora, for example, doesn't ship with it. So if we remove the toggle from the applet, then Fedora users would need to use dconf/gsettings to change it manually...

**Comment by vkareh** [comment]
another option is to add it to the mate-window-properties and change the text in the workspace-switcher one mentioning it's for mouse-scrolling only, that might be the easiest change of all

**Comment by knopancakes** [comment]
Yes, with all that in mind, I think adding mate-window-properties is probably the cleanest solution. The duplicate settings in mate-tweak are handled okay right now, so adding more should be fine. We also avoid removing functionality for users on other systems by simply cleaning up the semantics of the text.

**Comment by muesli4** [comment]
As I've stated on the launchpad page already, the settings are not related in any way. But people don't read. (_And then write rude and childish personal messages to me._)

@vkareh The workspace switcher applet preferences dialog mentions that this setting is for the switcher (in fact there is a setting for each workspace switcher applet). So there could be scenarios where you want to have it enabled in one workspace switcher but not in the other. This is consistent with most other options there except the number of workspaces. In light of that, I don't see any reason for changing the behavior.

@knopancakes If the description is that bad make a proposal for a better toggle button text.

**Comment by lukefromdc** [comment]
Mate-tweak on Debian mostly works but lots of python errors and problems saving parts of panel configuration.

**Comment by knopancakes** [comment]
@muesli4 You're missing the point. We know the settings aren't technically related, we're saying that there should be some way of controlling them both in a GUI.

In my opinion as a user, the settings are actually related because they both affect the ways in which I navigate my workspaces, so it makes sense to have the ability to control it from somewhere. I don't care if that's the panel app or the settings menu. If you followed the thread, we pretty much reached the consensus that the settings menu made more sense.

For a better name you need three words. "Allow workspace wrap around with scroll wheel in switcher"

**Comment by muesli4** [comment]
> You're missing the point, baby.

@knopancakes Baby? Can we keep this discussion professional, please? (*I also do not wish to receive any more personal emails from you.*)

> We know the settings aren't technically related, we're saying that there should be some way of controlling them both in a GUI.

Well, a day ago you didn't and now you're speaking for everyone. Obviously a lot of people have no clue what's going on. How about you respond to the points I brought up? I invented, implemented and maintained this feature. I can't think of a person that's more qualified to discuss with.

Personally, what I like about MATE is the ability to configure everything to my own needs. Usability over simplicity (because then it would be Gnome 3).

**Comment by knopancakes** [comment]
@muesli4 Yeah, baby! It's a colloquialism. You must be mistakenly interpreting it as an insult because you sent me a rude personal email with the same title. And here I thought you were just being friendly...

If you want a professional discussion, don't start out by insulting people. If you don't want to receive personal emails, don't send your own.

I'm not speaking for everyone, I'm speaking for myself and for the others who have directly contributed to this conversation (vkareh).

I responded to every single point you brought up... maybe you didn't read because you were too busy repeating that the settings aren't related. Now I understand why you're so defensive about this, your ego is clearly overinflated.

At least this we can agree on.

**Comment by N0rbert** [comment]
I can't understand what is going on here. The linked LP 1828842 bug was reported by *Theodor (hoffer-schaefer.theo)*, but here we did not hear from him. I filed this bug here to inform upstream developers.

Instead we have non-pleasant discussion with rude and non-technical behavior... What was the final target?

**Comment by raveit65** [comment]
Can please stop speaking like white-trash people?
Next time i will lock topic.

**Comment by knopancakes** [comment]
@N0rbert hoffer-schaefer.theo is me

**Comment by muesli4** [comment]
I requested more information from another bug report because he wouldn't tell me his version and did not read what I was writing. That's when he started sending me emails via launchpad (to which I responded, which obviously was my mistake). Also, just look at the history of @knopancakes post, he intentionally added *baby* to provoke me (context). I just want to clarify this because I don't like being slandered.

I've wrote everything there is to write about the topic here:

https://github.com/mate-desktop/mate-panel/issues/969#issuecomment-492749826

> The workspace switcher applet preferences dialog mentions that this setting is for the switcher (in fact there is a setting for each workspace switcher applet). So there could be scenarios where you want to have it enabled in one workspace switcher but not in the other. This is consistent with most other options there except the number of workspaces. In light of that, I don't see any reason for changing the behavior.

That's basically what keeps this setting from being centralized. No response to that so far.

> I don't think we've reached a consensus, as some of us feel that the settings are not related in any way despite them both having "workspace" and "wrap" in the name and affected functionality.

@knopancakes Because those settings are related to the behavior of two different application and one of them (*workspace-switcher-applet*) can have several different instances with different settings. I see the following alternatives:

* Remove the setting per *workspace-switcher-applet* and use a centralized one. Loss of functionality.
* Keep everything as is and improve text and documentation.

As for **mate-tweak**: You can still access this setting and display it. However, I don't think we should feel pressured to change anything to the worse just to make it easier for an external tool.

**Comment by knopancakes** [comment]
@muesli4 It takes two to tango my man. I don't like being referred to as illiterate. I can read, my mind just works differently than yours. I'm asking for the text to be improved for numbskulls like myself who can read just fine but think in a different context.

> So there could be scenarios where you want to have it enabled in one workspace switcher but not in the other.

Understood, the mouse wheel setting cannot be centralized. The context of that setting was misunderstood, but now that we have clarity this is not what is being asked for.

> This is consistent with most other options there **except the number of workspaces**

The application is called worspace-switcher-applet, and it already has hooks to manipulate settings like the number of workspaces (which is technically part of a different application). 

Don't remove anything, but rather extend the functionality. I think it's fitting for the application to have control over such settings as **org.mate.Marco.WrapStyle**, because to me they have very similar context.

I have the following, and I see no way whatsoever to control **org.mate.Marco.WrapStyle**:
```
~> apt-cache policy mate-tweak
mate-tweak:
  Installed: 18.04.16-1
  Candidate: 18.04.16-1
  Version table:
 *** 18.04.16-1 500
        500 http://us.archive.ubuntu.com/ubuntu bionic/universe amd64 Packages
        500 http://us.archive.ubuntu.com/ubuntu bionic/universe i386 Packages
        100 /var/lib/dpkg/status

```

**Comment by raveit65** [comment]
mate-tweak is a third party tool from ubuntu-mate.
Why not using dconf-editor (GUI) like users in other distros like fedora?
Here you'll find all gsettings keys for Mate and it has a search function.
Or is this too difficult for you?

**Comment by knopancakes** [comment]
I know mate-tweak is a third party tool, I'm simply saying that you can't "access this setting and display it" through mate-tweak as some have claimed, so it's a useless suggestion.

dconf-editor is not difficult to use. 

The existing implementation of mate-workspace-switcher is ambiguous, and I'm simply suggesting improvement to the cohesiveness of the application and desktop as a whole.

**Comment by vkareh** [comment]
I'd be happy to review a pull request implementing either the feature request or the change in text.

If there's merit in the change, and other MATE core members approve, we might merge it.

This conversation is now taking time from actual development and decision making. I'm OK if we lock this thread.
